File: 03-RECURSION/climbing-stairs.php
<?php


# https://leetcode.com/problems/climbing-stairs/
class Solution
{
    public function climbStairs(int $n): int
    {
        if ($n <= 1) {
            return 1;
        }

        return $this->climbStairs($n - 1) + $this->climbStairs($n - 2);
    }

    public function climbStairsFaster(int $n): int
    {
        $memo = [];

        return $this->helper($n, $memo);
    }

    public function helper(int $n, array $memo): int
    {
        if ($n <= 1) {
            return 1;
        }

        if (! in_array($n, $memo)) {
            $memo[$n] = $this->helper($n - 1, $memo) + $this->helper($n - 2, $memo);
        }

        return $memo[$n];
    }
}

File: 03-RECURSION/Go/climbing-stairs.go
package main

// https://leetcode.com/problems/climbing-stairs/
func climbStairs(n int) int {
	memo := make(map[int]int, 45)
	return calc(n, memo)
}

func calc(n int, memo map[int]int) int {
	if n < 2 {
		return 1
	}

	if _, ok := memo[n]; !ok {
		memo[n] = calc(n-1, memo) + calc(n-2, memo)
	}

	return memo[n]
}
